### ISSUE TITLE
vocab.transform_to in spa connections

### ISSUE THREAD

**Issue by drasmuss** [nengo/nengo#696]
Was playing around with the associative memory, and noticed this effect.  Here's a stripped down script to illustrate:

``` python
with spa.SPA() as model:
    d = 8

    parent = model.get_default_vocab(d)
    for i in range(4):
        print chr(65 + i), parent[chr(65 + i)]

    child = parent.create_subset(["A", "B"])

    model.buffer = spa.Buffer(d)
    model.assoc = spa.AssociativeMemory(child)

    actions = spa.Actions("assoc = buffer")
    model.cortical = spa.Cortical(actions)

    for k in child.keys:
        print k, child[k]
    print parent.transform_to(child)
```

There's a couple things going on here.
1. The `child` vocabulary now contains all the vectors in `parent`, not just the subset it was initialized with.  This is because the `Cortical` module, when it creates a connection from `buffer` to `assoc`, calls `parent.transform_to(child)`, which adds all the vocabulary in `parent` to `child`.  It's not really obvious how or why that happens unless you go digging through the code, so it's one of those "gotchas" for new users.
2. The vectors that get added to `child` do not match the vectors in `parent`.  So in the above example, "A" and "B" would match, but "C" and "D" would not.  Again, probably unexpected to most people.
3. The above are just usability issues, but I think this also has an impact on performance.  This is because the output of `transform_to` is a matrix, which is being used as the transform between `buffer` and `assoc`.  So when these new vectors get added in, what that's doing is giving you a noisy transform matrix, when what you'd actually like is no transform at all (you just want to feed "A" and "B" directly from `buffer` to `assoc`).  Note that this is true even if all the vectors did exactly match in `parent` and `child` (you're still going to get an unwanted, noisy transform matrix).

However, I do not have the best grasp on the internal workings of `spa`, so it's very possible that I'm misunderstanding something here.  Or there's a better way to do this that doesn't run into this problem.  That's the main reason I'm posting this, to get @tcstewar's insight.

**Comment by tcstewar** [comment]
Yup, that's a horribly documented side effect of the transform_to system.  This is actually the purpose of the ```keys``` parameter in transform_to: you can indicate what keys to consider.  If you don't indicate what keys, it assumes you want the union of all keys in the two vocabularies.  Most of the time, that's what I've wanted, but it's certainly not all of the time.

And the biggest problem is that I can't think of a good way of indicating in ```spa``` what to do (i.e. whether to use all keys in both, or whether you want only the keys that are in both, or whether you want some other list of keys).  Here's one possibility that I've kinda played with, but I'm definitely not convinced by it:

```python
actions = spa.Actions("assoc = buffer[[A,B]]")
```

I could also see something like
```python
actions = spa.Actions("assoc = buffer.only(A, B)")
```

Or perhaps you could mark a vocabulary to not add new items

```
child = parent.create_subset(["A", "B"])
child.allow_new_items = False
```

Any other ideas?

**Comment by xchoo** [comment]
Just my 2cents:

I think part of the problem is how we treat subset vocabularies. Currently, when you create a subset of a vocabulary, it makes a new vocabulary object and manually adds the subset vectors to the new vocabulary. If we revamped the subsetting system such that it could identify what the parent vocabulary is (and not create a transform if the parent vocabularies are the same) we can circumvent  this issue?

**Comment by xchoo** [comment]
@tcstewar We could mark the vocabulary to not add new items, but from what I can tell, even with identical vocabularies, the ```transform_to``` function doesn't create an identity transform.

```python
vocab2 = Vocabulary(16)
vocab2.parse('A+B+C+D')
t = vocab2.transform_to(vocab2)
a2 = np.dot(vocab2['A'].v, t)
print vocab2['A']
print a2
```

If I am not mistaken, in the code snippet above, in the idea case, both ```vocab2['A'].v``` and ```a2``` should be identical, no?

**Comment by tcstewar** [comment]
> but from what I can tell, even with identical vocabularies, the transform_to function doesn't create an identity transform

Hmm...  that's an interesting special case.  We could check for that condition, but I'd never considered that a person would ever want to make two separate identical vocabularies.  Still, if we wanted to handle that, we can pretty easily check for in in ```transform_to```.  Although it's also not quite what might be expected , since that'd give you a transform that would also map from a new vector E into that same vector in vocab2, which may not be what people want.

**Comment by jgosmann** [comment]
Some of this is addressed with #1014.

**Comment by jgosmann** [comment]
I think, this is resolved by #1137 (if not feel free to reopen this issue).

With the changes introduced in #1137:

1. One has to be explicit about when and how to convert between vocabularies.
2. Running the code from the original post it doesn't seem that new vectors get added to `child`. Neither for a `translate` nor a `reinterpret`.
3. @xchoo highlighted that `transform_to` does not check whether an identity transform is possible. But if one really wants an identity transform `reinterpret` should now be used to convert between those vocabularies. Thus, I don't think we have to check for this case in `transform_to` (which is used by `translate`; we should probably rename `transform_to` to `translate_to` for consistency).
